server/server.go

```diff
@@ -777,6 +778,82 @@ func (r *rodeServer) UpdatePolicy(ctx context.Context, updatePolicyRequest *pb.U
 	return policy, nil
 }
 
+func (r *rodeServer) RegisterCollector(ctx context.Context, registerCollectorRequest *pb.RegisterCollectorRequest) (*pb.RegisterCollectorResponse, error) {
+	log := r.logger.Named("RegisterCollector")
+
+	if registerCollectorRequest.Id == "" {
+		return nil, createErrorWithCode(log, "collector ID is required", nil, codes.InvalidArgument)
+	}
+
+	if len(registerCollectorRequest.Notes) == 0 {
+		return &pb.RegisterCollectorResponse{}, nil
+	}
+
+	// build collection of notes that potentially need to be created
+	notesWithIds := make(map[string]*grafeas_proto.Note)
+	notesToCreate := make(map[string]*grafeas_proto.Note)
+	for _, note := range registerCollectorRequest.Notes {
+		noteId := buildNoteIdFromCollectorId(registerCollectorRequest.Id, note)
+
+		if _, ok := notesWithIds[noteId]; ok {
+			return nil, createErrorWithCode(log, "cannot use more than one note type when registering a collector", nil, codes.InvalidArgument)
+		}
+
+		notesWithIds[noteId] = note
+		notesToCreate[noteId] = note
+	}
+
+	log = log.With(zap.Any("notes", notesWithIds))
+
+	// find out which notes already exist
```

Comment:
In grafeas-elasticsearch `BatchCreateNotes` is already doing a similar check and returning a specific [error](https://github.com/rode/grafeas-elasticsearch/blob/5c2d66454947d8b8e1a3ecda33b40ae504c8763e/go/v1beta1/storage/elasticsearch.go#L603) when a note already exists, would it be possible to reuse that? 

It's likely this would still need to handle the case where that error is returned, since it's possible that multiple collectors of the same type could start around the same time.

Reply:
Yeah I originally wanted to do this, but the grafeas API layer isn't really compatible with our implementation of `BatchCreateNotes` in this way - if any errors occur, it won't return any of them.  https://github.com/grafeas/grafeas/blob/c188e86dabdab1e99f05772aee40940b1ca6e6e2/go/v1/api/note.go#L102-L106